### PR TITLE
[SPARK-42289][SQL] DS V2 pushdown could let JDBC dialect decide to push down offset and limit

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
Currently, DS V2 pushdown force push down `LIMIT` clause and `OFFSET` clause to underlying databases. The behavior is not correct. For example, Oracle and MsSQLServer doesn't support `LIMIT` clause. DS V2 pushdown could let JDBC dialect decide to push down `OFFSET` and `LIMIT`.

### Why are the changes needed?

1. If the underlying database don't support `LIMIT` clause or `OFFSET` clause, the corresponding dialect could rejects push down them.
2. Although some underlying database don't support `LIMIT` clause or `OFFSET` clause,  the corresponding dialect have a lot of trick to finish the same function.

### Does this PR introduce _any_ user-facing change?
'No'.
New feature.

### How was this patch tested?
New test cases.

### CODE REVIEW

**Comment by beliefer** [comment]
ping @huaxingao cc @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/jdbc/JDBCScanBuilder.scala [hunk 1]
```diff
@@ -135,15 +135,16 @@ case class JDBCScanBuilder(
   }
 
   override def pushLimit(limit: Int): Boolean = {
-    if (jdbcOptions.pushDownLimit) {
+    if (jdbcOptions.pushDownLimit && JdbcDialects.get(jdbcOptions.url).supportsLimit) {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/jdbc/JDBCScanBuilder.scala:138]: shall we create a dialect instance as a field of `JDBCScanBuilder`?

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala [hunk 2]
```diff
@@ -557,6 +557,20 @@ abstract class JdbcDialect extends Serializable with Logging {
...
+   * Note: Some build-in dialect supports LIMIT clause with some trick, please see:
+   * {@link OracleDialect.OracleSQLQueryBuilder} and
+   * {@link MsSqlServerDialect.MsSqlServerSQLQueryBuilder}.
+   */
+  def supportsLimit: Boolean = true
+
+  /**
+   * Returns ture if dialect supports OFFSET clause.
+   */
+  def supportsOffset: Boolean = true
```
Comment (huaxingao) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:572]: Do all the dialects support offset? If not, do we need to set this to false in the dialect?

BTW you still have this `TODO[SPARK-42289]: DS V2 pushdown could let JDBC dialect decide to push down offset` in `OracleDialect` and `MsSqlServerDialect`

Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:572]: Most databases can directly or indirectly support offset. so I think we should keep true here.

Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:572]: > BTW you still have this `TODO[SPARK-42289]: DS V2 pushdown could let JDBC dialect decide to push down offset` in `OracleDialect` and `MsSqlServerDialect`

Yes. It need another PR. This PR just open the API.

**Review by huaxingao — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/jdbc/JDBCScanBuilder.scala [hunk 3]
```diff
@@ -126,24 +126,23 @@ case class JDBCScanBuilder(
       upperBound: Double,
       withReplacement: Boolean,
       seed: Long): Boolean = {
-    if (jdbcOptions.pushDownTableSample &&
-      JdbcDialects.get(jdbcOptions.url).supportsTableSample) {
+    if (jdbcOptions.pushDownTableSample && dialect.supportsTableSample) {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/jdbc/JDBCScanBuilder.scala:129]: shall we make two more changes?
1. remove the JDBC option for pushdown (false by default AFAIK)
2. jdbc dialect should return false in these new pushdown APIs, to keep the old false by default behavior.

Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/jdbc/JDBCScanBuilder.scala:129]: OK. Let's do it in next PR.

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:572]: I think it's better to use false as the default value, and override it in builtin dialects in return true.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:572]: given that we will turn on the JDBC option to enable pushdown soon.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:572]: We can do it in the PR switching the default value of JDBC pushdown options.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by beliefer** [comment]
@cloud-fan @huaxingao Thank you!

**Pull request closed without merging**